### PR TITLE
Enforce 100% branch coverage

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: patchy/api.py [hunk 1]
```diff
@@ -275,14 +275,8 @@ def _process_function():
     six.exec_(new_code, func.__globals__, localz)
     new_func = localz['__patchy_freevars__']()
 
-    # Figure out how to get the Code object
-    if isinstance(new_func, (classmethod, staticmethod)):
```
Comment (graingert) [inline patchy/api.py]: so the code in the new_source is never actually decorated with a classmethod or staticmethod - and it doesn't need to be (it seems)

Comment (graingert) [reply patchy/api.py]: oh wait no - it's a python 2 thing

File: .gitignore [hunk 2]
```diff
@@ -1,44 +1,126 @@
...
 pip-log.txt
+pip-delete-this-directory.txt
 
 # Unit test / coverage reports
-.cache
+htmlcov/
+.tox/
+.nox/
 .coverage
-.coverage*
```
Comment (graingert) [inline .gitignore]: this was incorrectly ignoring  `.coveragerc`

File: tests/test_patch.py [hunk 3]
```diff
@@ -56,14 +56,25 @@ def test_patch_invalid():
...
 
-    bad_patch = """
+    bad_patch = """garbage
         """
     with pytest.raises(ValueError) as excinfo:
         patchy.patch(sample, bad_patch)
 
     msg = str(excinfo.value)
-    assert msg.startswith("Could not apply the patch to 'sample'.")
-    assert "Only garbage was found in the patch input."
```
Comment (graingert) [inline tests/test_patch.py]: woops, this assert didn't do anything

File: .coveragerc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+[report]
+exclude_lines=
+    pragma: no cover
+    pragma: no py${PYTEST_PYMAJVER} cover
```
Comment (adamchainz) [inline .coveragerc]: TIL, nice!

File: pytest.ini [hunk 5]
```diff
@@ -1,5 +1,6 @@
 [pytest]
 xfail_strict=true
 addopts = --cov=patchy
+          --cov-branch
```
Comment (adamchainz) [inline pytest.ini]: 👌

File: tests/test_patch.py (hunk 3 above)
Comment (adamchainz) [reply tests/test_patch.py]: Well spotted

**Pull request merged**